In [146]:
import os
os.getcwd()

'/Users/ashutoshsaini/eniac-discount-analysis/notebooks'

In [147]:
import pandas as pd
orders_df = pd.read_csv('../data/raw/orders.csv')

In [148]:
orders_orig = orders_df.copy()

In [149]:
orderlines_df = pd.read_csv('../data/raw/orderlines.csv')

In [150]:
products_df = pd.read_csv('../data/raw/products.csv')

In [151]:
brands_df = pd.read_csv('../data/raw/brands.csv')

In [152]:

orders_df.head(10)

order_id         created_date  total_paid      state
0    241319  2017-01-02 13:35:40       44.99  Cancelled
1    241423  2017-11-06 13:10:02      136.15  Completed
2    242832  2017-12-31 17:40:03       15.76  Completed
3    243330  2017-02-16 10:59:38       84.98  Completed
4    243784  2017-11-24 13:35:19      157.86  Cancelled
5    245275  2017-06-28 11:35:37      149.00  Completed
6    245595  2017-01-21 12:52:47      112.97  Completed
7    245851  2017-04-04 20:58:21       79.99    Pending
8    245941  2017-01-01 10:32:23      183.52  Completed
9    246018  2017-02-13 23:02:57      211.95  Completed

In [153]:
orders_df.nunique()

order_id        226909
created_date    224828
total_paid       31236
state                5
dtype: int64

In [154]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


In [155]:
orders_df.isnull().any()

order_id        False
created_date    False
total_paid       True
state           False
dtype: bool

In [156]:
orders_df.isnull().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [157]:
orders_df['order_id'].unique()

array([241319, 241423, 242832, ..., 527399, 527400, 527401])

In [158]:
state_mask = orders_df['state'] == 'Completed'
#state_mask
orders_df.loc[state_mask].nunique()

order_id        46605
created_date    46506
total_paid      12834
state               1
dtype: int64

In [159]:
#state_mask.isnull().sum 

#this is not helping me, my fault here was that I was using boolean on a Series and not a df.

In [160]:
completed_orders = orders_df.loc[state_mask]

completed_orders['total_paid'].isnull().sum()


np.int64(0)

In [161]:
#I would like to know the number of completed orders

state_mask.value_counts()

state
False    180304
True      46605
Name: count, dtype: int64

In [162]:
# I would now like to know the percentage of completed orders, just to make proportion of completed orders more visible

completed_orders_percentage = (state_mask.sum()/len(state_mask))*100

completed_orders_percentage

np.float64(20.53907072879436)

Only 20.54% of orders(completed orders) contribute to the total revenue of Eniac 

In [163]:
#I will now calculate the percenatge of pending and cancelled orders

pending_order_mask = orders_df['state'] == 'Pending'

pending_orders_count = (orders_df['state'] == 'Pending').sum()
percentage_of_pending_orders = (pending_orders_count/len(pending_order_mask))*100

percentage_of_pending_orders

np.float64(6.336901577284286)

In [164]:
cancelled_order_mask = orders_df['state'] == 'Cancelled'

cancelled_orders_count = (orders_df['state'] == 'Cancelled').sum()
percentage_of_cancelled_orders = (cancelled_orders_count/len(cancelled_order_mask))*100

percentage_of_cancelled_orders

np.float64(3.187621469399627)

# Now I will calculate the percentage share of Shopping Basket orders and Place Order state.

In [165]:
shopping_basket_order_mask = orders_df['state'] == 'Shopping Basket'

shopping_basket_order_mask_count = (orders_df['state'] == 'Shopping Basket').sum()
percentage_of_shopping_basket_order_mask = (shopping_basket_order_mask_count/len(shopping_basket_order_mask))*100

percentage_of_shopping_basket_order_mask

np.float64(51.9190512496199)

In [166]:
place_order_mask = orders_df['state'] == 'Place Order'

place_order_mask_count = (orders_df['state'] == 'Place Order').sum()
percentage_of_place_order_mask = (place_order_mask_count/len(place_order_mask))*100

percentage_of_place_order_mask

np.float64(18.017354974901835)

In [167]:
(orders_df['state'].value_counts(normalize=True) * 100)#.sum()

state
Shopping Basket    51.919051
Completed          20.539071
Place Order        18.017355
Pending             6.336902
Cancelled           3.187621
Name: proportion, dtype: float64

In [168]:
# sanity Check

total_state_masks_count = percentage_of_pending_orders + percentage_of_shopping_basket_order_mask + percentage_of_cancelled_orders + completed_orders_percentage + percentage_of_place_order_mask


total_state_masks_count

np.float64(100.00000000000001)

The Shopping Basket orders account for ~51.92% and Completed orders account for ~20.54%.

In [169]:
completed_orders = orders_df.loc[state_mask]

completed_orders_df = completed_orders.copy()

In [170]:
orderlines_df.head()

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   

                  date  
0  2017-01-01 00:07:19  
1  2017-01-01 00:19:45  
2  2017-01-01 00:20:57  
3  2017-01-01 00:51:40  
4  2017-01-01 01:06:38

In [171]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                desc  price promo_price  \
0       Aluminum support compatible with all MacBook  59.99     499.899   
1          USB ultrathin keyboard Apple Mac Spanish.     59     589.996   
2                             mouse Apple USB cable.     59     569.898   
3              IPhone dock and USB Cable Apple iPod.     25     229.997   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pr...  34.99       31.99   

   in_stock      type  
0         1      8696  
1         0  13855401  
2         0      1387  
3         0      1230  
4         1      1364

Now that I have decided my nect two dataframes